libherokubuildpack/src/buildpack_output/ansi_escape.rs

```diff
@@ -0,0 +1,115 @@
+/// Wraps each line in an ANSI escape sequence while preserving prior ANSI escape sequences.
+///
+/// ## Why does this exist?
+///
+/// When buildpack output is streamed to the user, each line is prefixed with `remote: ` by Git. Any colorization
+/// of text will apply to those prefixes which is not the desired behavior. This function colors lines of text
+/// while ensuring that styles are disabled at the end of each line.
+///
+/// ## Supports recursive colorization
+///
+/// Strings that are previously colorized will not be overridden by this function. For example, if a word is already
+/// colored yellow, that word will continue to be yellow.
+pub(crate) fn wrap_ansi_escape_each_line(ansi: &ANSI, body: impl AsRef<str>) -> String {
+    let ansi_escape = ansi.to_str();
+    body.as_ref()
+        .split('\n')
+        // If sub contents are colorized it will contain SUBCOLOR ... RESET. After the reset,
+        // ensure we change back to the current color
+        .map(|line| line.replace(RESET, &format!("{RESET}{ansi_escape}"))) // Handles nested color
+        // Set the main color for each line and reset after so we don't colorize `remote:` by accident
+        .map(|line| format!("{ansi_escape}{line}{RESET}"))
+        // The above logic causes redundant colors and resets, clean them up
+        .map(|line| line.replace(&format!("{ansi_escape}{ansi_escape}"), ansi_escape)) // Reduce useless color
+        .map(|line| line.replace(&format!("{ansi_escape}{RESET}"), "")) // Empty lines or where the nested color is at the end of the line
+        .collect::<Vec<String>>()
+        .join("\n")
+}
+
+const RESET: &str = "\x1B[0m";
+const RED: &str = "\x1B[0;31m";
+const YELLOW: &str = "\x1B[0;33m";
+const BOLD_CYAN: &str = "\x1B[1;36m";
+const BOLD_PURPLE: &str = "\x1B[1;35m";
+
+#[derive(Debug)]
+#[allow(clippy::upper_case_acronyms)]
+pub(crate) enum ANSI {
+    Red,
+    Yellow,
+    BoldCyan,
+    BoldPurple,
+}
```

Comment:
It's an internal interface that the end user never touches. We can refactor it at will, and I have no intention of opening this function up to end-user consumption. I don't think we need to block on it.

Reply:
> then I realized that would mean my method signature was lying because it says it takes an &str.

Made me smile and I _really appreciate_ this train of thought! :) And I do agree it is lying and if we're talking about perfection I'd love to see a `AnsiEscape` type that neatly encapsulates everything that can be in an ANSI escape sequence, allows for parsing, validation and nice testing. I actually very briefly tried this in the initial review and refactor cycle before moving on. I realised that designing and implementing a full `AnsiEscape` is a big task and IMO out of scope for this PR. If we really want this, we should consider using [termcolor](https://docs.rs/termcolor/latest/termcolor/) where it's already implemented for us.

For the implementation here, I don't think we get enough value for the added code. The old constants are still around (for testing) hinting that we didn't really abstract. The new `ANSI` type is just a collection of specific combinations that we happen to need for the current functionality. Since this only internally used and the new `ANSI` type only guards what can be passed into `inject_default_ansi_escape` (not adding actual abstraction for downstream code) going the pragmatic route and leaving it as `&str` seems preferable to me in this case.

Reply:
> It's an internal interface that the end user never touches. We can refactor it at will, and I have no intention of opening this function up to end-user consumption. I don't think we need to block on it.

The reason we're discussing this is because after we'd reviewed this implementation, you changed the implementation even though it wasn't something we'd requested in the review comments. Changing things after initial review is not necessarily an issue, but I think it's then fair for us to discuss the changes, and not surprising if we ask you to change things back if we think the changes weren't for the better. And therefore, the "I don't think we need to block on it" seems like a bit of an unfair framing of any delay here.

Re this being internal only - I agree - but I think that strengths the position of the original API being fine. That is, since it's not a public API, it doesn't need to defend against mis-use quite as much, so using `&str` is fine (given the trade-offs Manuel mentioned).

Reply:
My main point is that it's relatively orthogonal to the whole PR review **AT ALL**. it's an internal interface that does not leak to the end user. Focusing our time here, unless it's to strengthen the guarantee that it doesn't leak in the future is...I'll not say a waste of time, though perhaps not a great use of time. Feel free to change my mind.

Reply:
>  you changed the implementation even though it wasn't something we'd requested in the review comments

This is a response to a refactor by Manuel posed as a refactor rather than a comment. I originally did not expose this function even to the rest of the crate (It is `pub(crate)` here, but that's an oversite https://github.com/heroku/buildpacks-ruby/blob/dda4ede413fc3fe4d6d2f2f63f039c7c1e5cc5fd/commons/src/output/fmt.rs#L189). I think it was okay before (after visibility would have been adjusted), but I worry about accidental use in the future.

> If we really want this, we should consider using [termcolor](https://docs.rs/termcolor/latest/termcolor/) where it's already implemented for us.

I forgot we have a `log.rs`, I can delete that and we can shave off yet another dependency.